File: C/cpu.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <stdatomic.h>
#include <mach/mach_time.h>

#define N 10000000
#define BRANCH_N 100000000
#define POINTER_N 50000000
#define SIMD_N 50000000

static mach_timebase_info_data_t tb;

static inline uint64_t now_ns(void) {
    if (tb.denom == 0) mach_timebase_info(&tb);
    return (mach_absolute_time() * tb.numer) / tb.denom;
}

// 1. Integer arithmetic
static void bench_fibonacci(void) {
    volatile unsigned long result = 0;
    uint64_t t0 = now_ns();
    for (int j = 0; j < N; j++) {
        unsigned long a = 0, b = 1;
        for (int i = 0; i < 50; i++) {
            unsigned long c = a + b;
            a = b; b = c;
        }
        result += a;
    }
    uint64_t t1 = now_ns();
    double ms = (t1 - t0) / 1e6;
    printf("cpu_fibonacci: %.2f ms (%.0f ops/sec)\n", ms, N / (ms / 1000.0));
    if (result == 0) printf("");
}

static void bench_prime_sieve(void) {
    int limit = 10000000;
    char* sieve = calloc(limit + 1, 1);
    if (!sieve) return;
    volatile int count = 0;
    uint64_t t0 = now_ns();
    for (int i = 2; i <= limit; i++) {
        if (!sieve[i]) {
            count++;
            if ((long)i * i <= limit)
                for (int j = i * i; j <= limit; j += i)
                    sieve[j] = 1;
        }
    }
    uint64_t t1 = now_ns();
    double ms = (t1 - t0) / 1e6;
    printf("cpu_prime_sieve: %.2f ms (%d primes)\n", ms, count);
    free(sieve);
}

static void bench_popcount(void) {
    int* data = malloc(N * sizeof(int));
    for (int i = 0; i < N; i++) data[i] = rand();
    volatile long total = 0;
    uint64_t t0 = now_ns();
    for (int i = 0; i < N; i++)
        total += __builtin_popcount(data[i]);
    uint64_t t1 = now_ns();
    double ms = (t1 - t0) / 1e6;
    printf("cpu_popcount: %.2f ms (%.0f ops/sec)\n", ms, N / (ms / 1000.0));
    free(data);
    if (total == 0) printf("");
}

// 2. Branch prediction stress
static void bench_branch_prediction(void) {
    int* sorted = malloc(BRANCH_N * sizeof(int));
    int* unsorted = malloc(BRANCH_N * sizeof(int));
    for (int i = 0; i < BRANCH_N; i++) {
        sorted[i] = i;
        unsorted[i] = rand() % 100;
    }
    
    volatile long sum_sorted = 0;
    uint64_t t0 = now_ns();
    for (int i = 0; i < BRANCH_N; i++) {
        if (sorted[i] > 50000000) sum_sorted += sorted[i];
    }
    uint64_t t1 = now_ns();
    double ms_sorted = (t1 - t0) / 1e6;
    printf("cpu_branch_sorted: %.2f ms\n", ms_sorted);
    
    volatile long sum_unsorted = 0;
    t0 = now_ns();
    for (int i = 0; i < BRANCH_N; i++) {
        if (unsorted[i] > 50) sum_unsorted += unsorted[i];
    }
    t1 = now_ns();
    double ms_unsorted = (t1 - t0) / 1e6;
    printf("cpu_branch_unsorted: %.2f ms (%.1fx slower)\n", ms_unsorted, ms_unsorted / ms_sorted);
    
    free(sorted); free(unsorted);
    if (sum_sorted == 0 || sum_unsorted == 0) printf("");
}

// 3. Memory latency (pointer chasing)
static void bench_memory_latency(void) {
    int* indices = malloc(POINTER_N * sizeof(int));
    for (int i = 0; i < POINTER_N; i++) indices[i] = (i + 1) % POINTER_N;
    // Shuffle
    for (int i = POINTER_N - 1; i > 0; i--) {
        int j = rand() % (i + 1);
        int t = indices[i]; indices[i] = indices[j]; indices[j] = t;
    }
    
    volatile int pos = 0;
    uint64_t t0 = now_ns();
    for (int i = 0; i < POINTER_N; i++)
        pos = indices[pos];
    uint64_t t1 = now_ns();
    double ms = (t1 - t0) / 1e6;
    printf("cpu_memory_latency: %.2f ms (%.0f traversals/sec)\n", ms, POINTER_N / (ms / 1000.0));
    free(indices);
    if (pos == 0) printf("");
}

// 4. NEON SIMD vector ops
#ifdef __ARM_NEON
#include <arm_neon.h>
static void bench_simd(void) {
    float* a = aligned_alloc(16, SIMD_N * sizeof(float));
    float* b = aligned_alloc(16, SIMD_N * sizeof(float));
    float* c = aligned_alloc(16, SIMD_N * sizeof(float));
    for (int i = 0; i < SIMD_N; i++) { a[i] = i * 1.0f; b[i] = (i % 100) * 1.0f; }
    
    volatile float result = 0;
    uint64_t t0 = now_ns();
    for (int i = 0; i < SIMD_N; i += 4) {
        float32x4_t va = vld1q_f32(a + i);
        float32x4_t vb = vld1q_f32(b + i);
        float32x4_t vc = vaddq_f32(va, vb);
        vc = vmulq_f32(vc, va);
        vst1q_f32(c + i, vc);
    }
    uint64_t t1 = now_ns();
    double ms = (t1 - t0) / 1e6;
    // Sum to prevent DCE
    for (int i = 0; i < 1000; i++) result += c[i * 1000];
    printf("cpu_simd_neon: %.2f ms (%.0f ops/sec)\n", ms, (SIMD_N / 4) / (ms / 1000.0));
    free(a); free(b); free(c);
    if (result == 0) printf("");
}
#else
static void bench_simd(void) {
    float* a = malloc(SIMD_N * sizeof(float));
    float* b = malloc(SIMD_N * sizeof(float));
    float* c = malloc(SIMD_N * sizeof(float));
    for (int i = 0; i < SIMD_N; i++) { a[i] = i * 1.0f; b[i] = (i % 100) * 1.0f; }
    
    volatile float result = 0;
    uint64_t t0 = now_ns();
    for (int i = 0; i < SIMD_N; i++)
        c[i] = (a[i] + b[i]) * a[i];
    uint64_t t1 = now_ns();
    double ms = (t1 - t0) / 1e6;
    for (int i = 0; i < 1000; i++) result += c[i * 1000];
    printf("cpu_simd_scalar: %.2f ms (%.0f ops/sec)\n", ms, SIMD_N / (ms / 1000.0));
    free(a); free(b); free(c);
    if (result == 0) printf("");
}
#endif

// 5. Atomic operations
static void bench_atomic(void) {
    atomic_int atomic_val = 0;
    uint64_t t0 = now_ns();
    for (int i = 0; i < N; i++)
        atomic_fetch_add(&atomic_val, 1);
    uint64_t t1 = now_ns();
    double ms_atomic = (t1 - t0) / 1e6;
    printf("cpu_atomic_add: %.2f ms (%.0f ops/sec)\n", ms_atomic, N / (ms_atomic / 1000.0));
    
    volatile int regular_val = 0;
    t0 = now_ns();
    for (int i = 0; i < N; i++)
        regular_val++;
    t1 = now_ns();
    double ms_reg = (t1 - t0) / 1e6;
    printf("cpu_nonatomic_add: %.2f ms (%.0f ops/sec, %.1fx faster)\n", ms_reg, N / (ms_reg / 1000.0), ms_atomic / ms_reg);
    
    if (atomic_val == 0 || regular_val == 0) printf("");
}

int main(void) {
    printf("C CPU Benchmark\n");
    printf("Apple M4 — 10 cores (4P+6E), 16GB unified\n\n");
    
    printf("--- Integer Arithmetic ---\n");
    bench_fibonacci();
    bench_prime_sieve();
    bench_popcount();
    
    printf("\n--- Branch Prediction ---\n");
    bench_branch_prediction();
    
    printf("\n--- Memory Latency ---\n");
    bench_memory_latency();
    
    printf("\n--- SIMD Vectorization ---\n");
    bench_simd();
    
    printf("\n--- Atomic Operations ---\n");
    bench_atomic();
    
    return 0;
}

File: Rust/cpu.rs
use std::time::Instant;

const N: usize = 10_000_000;
const BRANCH_N: usize = 100_000_000;
const POINTER_N: usize = 50_000_000;
const SIMD_N: usize = 50_000_000;

fn fmt_ms(d: std::time::Duration) -> f64 {
    d.as_secs_f64() * 1000.0
}

// 1. Integer arithmetic
fn bench_fibonacci() {
    let mut results = Vec::with_capacity(N);
    let t0 = Instant::now();
    for j in 0..N {
        let (mut a, mut b) = (0u64, 1u64);
        for _ in 0..50 {
            let c = a.wrapping_add(b);
            a = b; b = c;
        }
        results.push(a ^ (j as u64));  // prevent DCE
    }
    let ms = fmt_ms(t0.elapsed());
    let sum: u64 = results.iter().sum();
    println!("cpu_fibonacci: {:.2} ms ({:.0} ops/sec)", ms, N as f64 / (ms / 1000.0));
    std::hint::black_box(sum);
    results.truncate(0);
}

fn bench_prime_sieve() {
    let limit = 10_000_000;
    let mut sieve = vec![false; limit + 1];
    let mut count = 0u64;
    let t0 = Instant::now();
    for i in 2..=limit {
        if !sieve[i] {
            count += 1;
            if (i as u64).wrapping_mul(i as u64) <= limit as u64 {
                let mut j = i.wrapping_mul(i);
                while j <= limit {
                    sieve[j] = true;
                    j += i;
                }
            }
        }
    }
    let ms = fmt_ms(t0.elapsed());
    println!("cpu_prime_sieve: {:.2} ms ({} primes)", ms, count);
}

fn bench_popcount() {
    let data: Vec<u32> = (0..N as u32).map(|i| i.wrapping_mul(0x9e3779b9)).collect();
    let mut total = 0u64;
    let t0 = Instant::now();
    for &x in &data {
        total += x.count_ones() as u64;
    }
    let ms = fmt_ms(t0.elapsed());
    println!("cpu_popcount: {:.2} ms ({:.0} ops/sec)", ms, N as f64 / (ms / 1000.0));
    std::hint::black_box(total);
}

// 2. Branch prediction
fn bench_branch_prediction() {
    use rand::Rng;
    let mut rng = rand::thread_rng();
    let unsorted: Vec<i64> = (0..BRANCH_N).map(|_| rng.gen_range(0..100)).collect();
    let mut sorted = unsorted.clone();
    sorted.sort();

    let t0 = Instant::now();
    let mut sum_unsorted: i64 = 0;
    for &x in &unsorted {
        if x > 50 { sum_unsorted += x; }
    }
    let ms_unsorted = fmt_ms(t0.elapsed());
    println!("cpu_branch_unsorted: {:.2} ms", ms_unsorted);

    let t0 = Instant::now();
    let mut sum_sorted: i64 = 0;
    for &x in &sorted {
        if x > 50 { sum_sorted += x; }
    }
    let ms_sorted = fmt_ms(t0.elapsed());
    println!("cpu_branch_sorted: {:.2} ms ({:.1}x faster)", ms_sorted, ms_unsorted / ms_sorted);

    std::hint::black_box((sum_unsorted, sum_sorted));
}

// 3. Memory latency (pointer chasing)
fn bench_memory_latency() {
    use rand::Rng;
    let mut rng = rand::thread_rng();
    let mut indices: Vec<usize> = (0..POINTER_N).collect();
    // Fisher-Yates shuffle
    for i in (1..POINTER_N).rev() {
        let j = rng.gen_range(0..=i);
        indices.swap(i, j);
    }

    let t0 = Instant::now();
    let mut pos = 0usize;
    for _ in 0..POINTER_N {
        pos = indices[pos];
    }
    let ms = fmt_ms(t0.elapsed());
    println!("cpu_memory_latency: {:.2} ms ({:.0} traversals/sec)", ms, POINTER_N as f64 / (ms / 1000.0));
    std::hint::black_box(pos);
}

// 4. SIMD — use explicit NEON on aarch64
#[cfg(target_arch = "aarch64")]
fn bench_simd() {
    use std::arch::aarch64::*;
    let mut a = vec![0.0f32; SIMD_N];
    let mut b = vec![0.0f32; SIMD_N];
    let mut c = vec![0.0f32; SIMD_N];
    for i in 0..SIMD_N { a[i] = i as f32; b[i] = (i % 100) as f32; }

    let t0 = Instant::now();
    unsafe {
        for i in (0..SIMD_N).step_by(4) {
            let va = vld1q_f32(a.as_ptr().add(i));
            let vb = vld1q_f32(b.as_ptr().add(i));
            let vc = vaddq_f32(va, vb);
            let vd = vmulq_f32(vc, va);
            vst1q_f32(c.as_mut_ptr().add(i), vd);
        }
    }
    let ms = fmt_ms(t0.elapsed());
    let mut total = 0.0f32;
    for i in (0..1000) { total += c[i * 1000]; }
    println!("cpu_simd_neon: {:.2} ms ({:.0} ops/sec)", ms, (SIMD_N as f64 / 4.0) / (ms / 1000.0));
    std::hint::black_box(total);
}

#[cfg(not(target_arch = "aarch64"))]
fn bench_simd() {
    let a: Vec<f32> = (0..SIMD_N).map(|i| i as f32).collect();
    let b: Vec<f32> = (0..SIMD_N).map(|i| (i % 100) as f32).collect();
    let mut c = vec![0.0f32; SIMD_N];

    let t0 = Instant::now();
    for i in 0..SIMD_N {
        c[i] = (a[i] + b[i]) * a[i];
    }
    let ms = fmt_ms(t0.elapsed());
    let mut total = 0.0f32;
    for i in (0..1000) { total += c[i * 1000]; }
    println!("cpu_simd_scalar: {:.2} ms ({:.0} ops/sec)", ms, SIMD_N as f64 / (ms / 1000.0));
    std::hint::black_box(total);
}

// 5. Atomic
fn bench_atomic() {
    use std::sync::atomic::{AtomicI64, Ordering};
    let atomic_val = AtomicI64::new(0);
    let t0 = Instant::now();
    for i in 0..N {
        atomic_val.fetch_add(std::hint::black_box(1), Ordering::Relaxed);
        std::hint::black_box(i);
    }
    let ms_atomic = fmt_ms(t0.elapsed());
    let atomic_final = atomic_val.load(Ordering::Relaxed);
    println!("cpu_atomic_add: {:.2} ms ({:.0} ops/sec)", ms_atomic, N as f64 / (ms_atomic / 1000.0));

    let mut regular_val = 0i64;
    let t0 = Instant::now();
    for i in 0..N {
        regular_val = regular_val.wrapping_add(1);
        std::hint::black_box(regular_val);
        std::hint::black_box(i);
    }
    let ms_reg = fmt_ms(t0.elapsed());
    println!("cpu_nonatomic_add: {:.2} ms ({:.0} ops/sec, {:.1}x faster)", ms_reg, N as f64 / (ms_reg / 1000.0), ms_atomic / ms_reg);

    std::hint::black_box((atomic_final, regular_val));
}

fn main() {
    println!("Rust CPU Benchmark");
    println!("Apple M4 — 10 cores (4P+6E), 16GB unified\n");

    println!("--- Integer Arithmetic ---");
    bench_fibonacci();
    bench_prime_sieve();
    bench_popcount();

    println!("\n--- Branch Prediction ---");
    bench_branch_prediction();

    println!("\n--- Memory Latency ---");
    bench_memory_latency();

    println!("\n--- SIMD Vectorization ---");
    bench_simd();

    println!("\n--- Atomic Operations ---");
    bench_atomic();
}
